### DISCUSSION TITLE
Deep Graph Infomax without node features

### DISCUSSION THREAD

**Discussion by naman-ali** [stellargraph/stellargraph#1855] [General] [answered]
Since one of the main parts of the DPI model relies on corrupting the graph, does it improve the results for non-attributed graphs (graphs without node features)?

_ **Corruption procedure**: The corruption procedure changes the true input graph G into the mutated version H = C(G). The paper suggests randomly shuffling the node features among the nodes: H has the same edges as G, but the features associated with each node differ._

**Comment by huonw** [comment, marked as answer]
Hi, thanks for asking a question. I think one could devise corruption procedures that work for graphs without attributes (for instance, many non-attribute methods are equivalent to using a one-hot encoded version of the node index, which could be regarded as attributes to shuffle), but the default ones built into StellarGraph are focused on graphs with attributes.

**Reply by naman-ali** [reply]
Thanks for the reply. Right, I found the idea of using one-hot encoded in one of your demos. 

I know what is one-hot encoding, but I am confused about using it for a node.

For example, assume my graph has 2 types of edges, and there are 100 nodes in total, so the attribute for each node is a vector of 200 dimensions, and if there is an edge I set the value to 1, and if not exist to 0. Is this what you mean?

**Reply by huonw** [reply]
Ah, good question! In this case, I'm meaning encoding the node itself. For your example, the attribute for each node would be 100 dimensions, and the first node's attributes would be `[1, 0, 0, ...]`, second node `[0, 1, 0, ..., 0]`, third node `[0, 0, 1, ..., 0]` and so on until the last node `[0, 0, 0, ..., 1]`. Does that make sense?

This direct encoding is unfortunately very inefficient, because node features take up `number of nodes^2` space, so it becomes intractable for graphs that are too large, but it may at least allow you to progress. 

(This topic of discussion is tangentially related to https://github.com/stellargraph/stellargraph/issues/1762 so I'm mentioning it here as a reference/reminder for me, for future.)

**Reply by naman-ali** [reply]
Yes, make sense. Thank you for the help.
